File: importer.js
/**
 * importer.js — Convert JMdict XML to a SQLite database.
 * 
 * Usage:
 *     node importer.js --xml data/JMdict --db data/jdict.db
 */

const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');
const { XMLParser } = require('fast-xml-parser');

function stripEntity(text) {
  if (typeof text !== 'string') return '';
  if (text.startsWith('&') && text.endsWith(';')) {
    return text.slice(1, -1);
  }
  return text;
}

function createSchema(db) {
  db.exec(`
    PRAGMA journal_mode=WAL;
    PRAGMA synchronous=NORMAL;

    CREATE TABLE IF NOT EXISTS entries (
      seq         INTEGER PRIMARY KEY,
      kanji_json  TEXT,
      kana_json   TEXT,
      senses_json TEXT,
      jlpt        TEXT
    );

    CREATE TABLE IF NOT EXISTS entries_text (
      seq     INTEGER PRIMARY KEY,
      kanji   TEXT NOT NULL DEFAULT '',
      kana    TEXT NOT NULL DEFAULT '',
      glosses TEXT NOT NULL DEFAULT ''
    );

    CREATE VIRTUAL TABLE IF NOT EXISTS entries_fts USING fts5(
      glosses,
      content='entries_text',
      content_rowid='seq'
    );
  `);
}

function parseEntry(entry) {
  const seq = parseInt(entry.ent_seq, 10) || 0;

  const kanjiList = [].concat(entry.k_ele || []).map(k => ({
    keb: k.keb || '',
    ke_inf: [].concat(k.ke_inf || []).map(stripEntity),
    ke_pri: [].concat(k.ke_pri || []),
  }));

  const kanaList = [].concat(entry.r_ele || []).map(r => ({
    reb: r.reb || '',
    re_nokanji: r.re_nokanji !== undefined,
    re_inf: [].concat(r.re_inf || []).map(stripEntity),
    re_pri: [].concat(r.re_pri || []),
    re_restr: [].concat(r.re_restr || []),
  }));

  const senses = [].concat(entry.sense || []).map(s => {
    const lsource = [].concat(s.lsource || []).map(ls => ({
      lang: ls['@_xml:lang'] || 'eng',
      text: ls['#text'] || ls || '',
      ls_type: ls['@_ls_type'] || '',
      ls_wasei: ls['@_ls_wasei'] || '',
    }));

    const glosses = [].concat(s.gloss || []).map(g => {
      let text, lang, g_type;
      if (typeof g === 'object') {
        text = g['#text'] || '';
        lang = g['@_xml:lang'] || 'eng';
        g_type = g['@_g_type'] || '';
      } else {
        text = g || '';
        lang = 'eng';
        g_type = '';
      }
      return { lang, text, g_type };
    });

    return {
      pos: [].concat(s.pos || []).map(stripEntity),
      misc: [].concat(s.misc || []).map(stripEntity),
      field: [].concat(s.field || []).map(stripEntity),
      dial: [].concat(s.dial || []).map(stripEntity),
      stagk: [].concat(s.stagk || []),
      stagr: [].concat(s.stagr || []),
      xref: [].concat(s.xref || []),
      ant: [].concat(s.ant || []),
      lsource,
      glosses,
      s_inf: [].concat(s.s_inf || []),
    };
  });

  return { seq, kanji: kanjiList, kana: kanaList, senses };
}

function importJmdict(xmlPath, dbPath) {
  console.log(`Opening ${xmlPath} ...`);
  const t0 = Date.now();

  const db = new Database(dbPath);
  createSchema(db);

  console.log("Reading XML (stripping DOCTYPE entities) ...");
  let raw = fs.readFileSync(xmlPath);

  // Strip DOCTYPE block (similar to Python logic)
  const doctypeEnd = raw.indexOf(']>');
  if (doctypeEnd !== -1) {
    const xmlDeclEnd = raw.indexOf('?>');
    if (xmlDeclEnd !== -1) {
      const header = raw.slice(0, xmlDeclEnd + 2);
      const body = raw.slice(doctypeEnd + 2);
      raw = Buffer.concat([header, body]);
    } else {
      raw = raw.slice(doctypeEnd + 2);
    }
  }

  // Replace XML entity refs (&v1; etc.) with their names as plain text
  let rawStr = raw.toString('utf8');
  rawStr = rawStr.replace(/&([A-Za-z0-9_-]+);/g, '$1');

  console.log("Parsing XML ...");
  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: "@_",
    allowBooleanAttributes: true,
  });
  const root = parser.parse(rawStr);

  const entries = root.JMdict.entry || [];
  console.log(`Found ${entries.length} entries. Importing...`);

  const insertEntry = db.prepare('INSERT OR REPLACE INTO entries VALUES (?,?,?,?,?)');
  const insertText = db.prepare('INSERT OR REPLACE INTO entries_text VALUES (?,?,?,?)');
  const insertFts = db.prepare('INSERT INTO entries_fts(rowid, glosses) VALUES (?,?)');

  const batchSize = 5000;
  let total = 0;

  db.transaction(() => {
    for (const entryObj of entries) {
      const e = parseEntry(entryObj);

      const kanjiText = e.kanji.map(k => k.keb).join(' ');
      const kanaText = e.kana.map(k => k.reb).join(' ');
      const glossText = e.senses
        .flatMap(s => s.glosses)
        .filter(g => g.lang === 'eng')
        .map(g => g.text)
        .join(' ');

      insertEntry.run(
        e.seq,
        JSON.stringify(e.kanji),
        JSON.stringify(e.kana),
        JSON.stringify(e.senses),
        null // jlpt
      );
      insertText.run(e.seq, kanjiText, kanaText, glossText);
      insertFts.run(e.seq, glossText);

      total++;
      if (total % batchSize === 0) {
        process.stdout.write(`  ... ${total.toLocaleString()} entries imported\r`);
      }
    }
  })();

  console.log("\nBuilding indexes ...");
  db.exec(`
    CREATE INDEX IF NOT EXISTS idx_text_kanji ON entries_text(kanji);
    CREATE INDEX IF NOT EXISTS idx_text_kana  ON entries_text(kana);
  `);

  console.log(`Done. ${total.toLocaleString()} entries imported in ${((Date.now() - t0) / 1000).toFixed(1)}s -> ${dbPath}`);
  db.close();
}

const args = require('minimist')(process.argv.slice(2));
const xmlPath = args.xml || 'data/JMdict';
const dbPath = args.db || 'data/jdict.db';

if (!fs.existsSync(xmlPath)) {
  console.error(`ERROR: XML file not found: ${xmlPath}`);
  process.exit(1);
}

if (!fs.existsSync(path.dirname(dbPath))) {
  fs.mkdirSync(path.dirname(dbPath), { recursive: true });
}

importJmdict(xmlPath, dbPath);

File: importer.py
"""
importer.py — Convert JMdict XML to a SQLite database.

Usage:
    python importer.py --xml data/JMdict --db data/jdict.db

The JMdict XML file can be downloaded from:
    https://www.edrdg.org/wiki/index.php/JMdict-EDICT_Dictionary_Project
    Direct: http://ftp.edrdg.org/pub/Nihongo/JMdict.gz  (then gunzip)
"""

import argparse
import sqlite3
import sys
import time
import xml.etree.ElementTree as ET
from pathlib import Path


def strip_entity(text: str) -> str:
    """'&v1;' -> 'v1'"""
    if text and text.startswith("&") and text.endswith(";"):
        return text[1:-1]
    return text or ""


def create_schema(conn: sqlite3.Connection):
    conn.executescript("""
        PRAGMA journal_mode=WAL;
        PRAGMA synchronous=NORMAL;

        CREATE TABLE IF NOT EXISTS entries (
            seq         INTEGER PRIMARY KEY,
            kanji_json  TEXT,
            kana_json   TEXT,
            senses_json TEXT,
            jlpt        TEXT
        );

        -- Plain text table for search (kanji/kana LIKE + FTS glosses)
        CREATE TABLE IF NOT EXISTS entries_text (
            seq     INTEGER PRIMARY KEY,
            kanji   TEXT NOT NULL DEFAULT '',
            kana    TEXT NOT NULL DEFAULT '',
            glosses TEXT NOT NULL DEFAULT ''
        );

        -- FTS5 over entries_text for English gloss search
        CREATE VIRTUAL TABLE IF NOT EXISTS entries_fts USING fts5(
            glosses,
            content='entries_text',
            content_rowid='seq'
        );
    """)
    conn.commit()


def parse_entry(entry_el) -> dict:
    import json

    seq = int(entry_el.findtext("ent_seq", "0"))

    kanji_list = []
    for k_el in entry_el.findall("k_ele"):
        kanji_list.append({
            "keb":    k_el.findtext("keb", ""),
            "ke_inf": [strip_entity(e.text) for e in k_el.findall("ke_inf")],
            "ke_pri": [e.text for e in k_el.findall("ke_pri")],
        })

    kana_list = []
    for r_el in entry_el.findall("r_ele"):
        kana_list.append({
            "reb":        r_el.findtext("reb", ""),
            "re_nokanji": r_el.find("re_nokanji") is not None,
            "re_inf":     [strip_entity(e.text) for e in r_el.findall("re_inf")],
            "re_pri":     [e.text for e in r_el.findall("re_pri")],
            "re_restr":   [e.text for e in r_el.findall("re_restr")],
        })

    senses = []
    for s_el in entry_el.findall("sense"):
        pos   = [strip_entity(e.text) for e in s_el.findall("pos")]
        misc  = [strip_entity(e.text) for e in s_el.findall("misc")]
        field = [strip_entity(e.text) for e in s_el.findall("field")]
        dial  = [strip_entity(e.text) for e in s_el.findall("dial")]
        stagk = [e.text for e in s_el.findall("stagk")]
        stagr = [e.text for e in s_el.findall("stagr")]
        xref  = [e.text for e in s_el.findall("xref")]
        ant   = [e.text for e in s_el.findall("ant")]
        lsource = []
        for ls in s_el.findall("lsource"):
            lsource.append({
                "lang": ls.get("{http://www.w3.org/XML/1998/namespace}lang", "eng"),
                "text": ls.text or "",
                "ls_type": ls.get("ls_type", ""),
                "ls_wasei": ls.get("ls_wasei", ""),
            })
        glosses = []
        for g in s_el.findall("gloss"):
            lang = g.get("{http://www.w3.org/XML/1998/namespace}lang", "eng")
            glosses.append({"lang": lang, "text": g.text or "", "g_type": g.get("g_type", "")})
        s_inf = [e.text for e in s_el.findall("s_inf")]

        senses.append({
            "pos": pos, "misc": misc, "field": field, "dial": dial,
            "stagk": stagk, "stagr": stagr, "xref": xref, "ant": ant,
            "lsource": lsource, "glosses": glosses, "s_inf": s_inf,
        })

    return {
        "seq": seq,
        "kanji": kanji_list,
        "kana": kana_list,
        "senses": senses,
    }


def import_jmdict(xml_path: Path, db_path: Path):
    import json, re

    print(f"Opening {xml_path} ...")
    t0 = time.time()

    conn = sqlite3.connect(db_path)
    create_schema(conn)

    print("Reading XML (stripping DOCTYPE entities) ...")
    raw = xml_path.read_bytes()

    # Strip DOCTYPE block
    doctype_end = raw.find(b"]>")
    if doctype_end != -1:
        xml_decl_end = raw.find(b"?>")
        if xml_decl_end != -1:
            header = raw[:xml_decl_end + 2]
            body = raw[doctype_end + 2:]
            raw = header + body
        else:
            raw = raw[doctype_end + 2:]

    # Replace XML entity refs (&v1; etc.) with their names as plain text
    raw = re.sub(rb"&([A-Za-z0-9_-]+);", rb"\1", raw)

    print("Parsing XML ...")
    root = ET.fromstring(raw)

    entries      = []
    text_rows    = []
    fts_rows     = []
    batch_size   = 5000
    total        = 0

    for entry_el in root.iter("entry"):
        e = parse_entry(entry_el)

        kanji_text  = " ".join(k["keb"] for k in e["kanji"])
        kana_text   = " ".join(k["reb"] for k in e["kana"])
        gloss_text  = " ".join(
            g["text"] for s in e["senses"] for g in s["glosses"] if g["lang"] == "eng"
        )

        entries.append((
            e["seq"],
            json.dumps(e["kanji"],  ensure_ascii=False),
            json.dumps(e["kana"],   ensure_ascii=False),
            json.dumps(e["senses"], ensure_ascii=False),
            None,  # jlpt
        ))
        text_rows.append((e["seq"], kanji_text, kana_text, gloss_text))
        fts_rows.append((e["seq"], gloss_text))
        total += 1

        if len(entries) >= batch_size:
            conn.executemany("INSERT OR REPLACE INTO entries VALUES (?,?,?,?,?)", entries)
            conn.executemany("INSERT OR REPLACE INTO entries_text VALUES (?,?,?,?)", text_rows)
            conn.executemany("INSERT INTO entries_fts(rowid, glosses) VALUES (?,?)", fts_rows)
            conn.commit()
            entries.clear(); text_rows.clear(); fts_rows.clear()
            print(f"  ... {total:,} entries imported", end="\r")

    if entries:
        conn.executemany("INSERT OR REPLACE INTO entries VALUES (?,?,?,?,?)", entries)
        conn.executemany("INSERT OR REPLACE INTO entries_text VALUES (?,?,?,?)", text_rows)
        conn.executemany("INSERT INTO entries_fts(rowid, glosses) VALUES (?,?)", fts_rows)
        conn.commit()

    # Index entries_text for fast LIKE searches
    print("\nBuilding indexes ...")
    conn.executescript("""
        CREATE INDEX IF NOT EXISTS idx_text_kanji ON entries_text(kanji);
        CREATE INDEX IF NOT EXISTS idx_text_kana  ON entries_text(kana);
    """)
    conn.commit()

    print(f"Done. {total:,} entries imported in {time.time()-t0:.1f}s -> {db_path}")
    conn.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Import JMdict XML into SQLite")
    parser.add_argument("--xml", default="data/JMdict",   help="Path to JMdict XML file")
    parser.add_argument("--db",  default="data/jdict.db", help="Output SQLite path")
    args = parser.parse_args()

    xml_path = Path(args.xml)
    db_path  = Path(args.db)

    if not xml_path.exists():
        print(f"ERROR: XML file not found: {xml_path}", file=sys.stderr)
        print("Download JMdict from: http://ftp.edrdg.org/pub/Nihongo/JMdict.gz", file=sys.stderr)
        sys.exit(1)

    db_path.parent.mkdir(parents=True, exist_ok=True)
    import_jmdict(xml_path, db_path)
